cmd/hooks/prestop.go

```diff
@@ -0,0 +1,175 @@
+	node, err := clientset.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("fetchNode: failed to retrieve node information: %w", err)
+	}
+	return node, nil
+}
+
+// isNodeBeingDrained returns true if node resource has a known drain/eviction taint.
+func isNodeBeingDrained(node *v1.Node) bool {
+	for _, taint := range node.Spec.Taints {
+		if _, isDrainTaint := drainTaints[taint.Key]; isDrainTaint {
+			return true
+		}
+	}
+	return false
+}
+
+func waitForPodShutdowns(clientset kubernetes.Interface, nodeName string) error {
+	allVolumesUnmounted := make(chan struct{})
+
+	factory := informers.NewSharedInformerFactoryWithOptions(clientset, 0,
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.FieldSelector = fmt.Sprintf("spec.nodeName=%s", nodeName)
+		}))
+	informer := factory.Core().V1().Pods().Informer()
+
+	_, err := informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			klog.V(5).InfoS("DeleteFunc: Pod deleted", "node", nodeName)
+			if err := checkActivePods(clientset, nodeName, allVolumesUnmounted); err != nil {
+				klog.ErrorS(err, "DeleteFunc: error checking active pods on the node")
+			}
+
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			klog.V(5).InfoS("UpdateFunc: Pod updated", "node", nodeName)
+			if err := checkActivePods(clientset, nodeName, allVolumesUnmounted); err != nil {
+				klog.ErrorS(err, "UpdateFunc: error checking active pods on the node")
+			}
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to add event handler to Node informer: %w", err)
+	}
+
+	go informer.Run(allVolumesUnmounted)
+
+	if err := checkActivePods(clientset, nodeName, allVolumesUnmounted); err != nil {
+		klog.ErrorS(err, "waitForPodShutdowns: error checking active pods on the node")
+	}
+
+	<-allVolumesUnmounted
+	klog.InfoS("waitForPodShutdowns: finished waiting for active pods on the node. preStopHook completed")
+	return nil
+}
+
+func checkActivePods(clientset kubernetes.Interface, nodeName string, allVolumesUnmounted chan struct{}) error {
+	podList, err := clientset.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
+	})
+	if err != nil {
+		klog.V(2).ErrorS(err, "failed to get podList")
+		return nil
+	}
+	for _, pod := range podList.Items {
+		// Temporary workaround until FieldSelector filters properly: https://github.com/kubernetes/client-go/issues/1350
+		if pod.Spec.NodeName == nodeName {
+			for _, vol := range pod.Spec.Volumes {
+
+				if vol.PersistentVolumeClaim == nil {
+					continue
+				}
+
+				pvcName := vol.PersistentVolumeClaim.ClaimName
+				pvc, err := clientset.CoreV1().PersistentVolumeClaims(pod.Namespace).Get(context.Background(), pvcName, metav1.GetOptions{})
+				if err != nil {
+					klog.V(2).ErrorS(err, "failed to get pvc", "pvc", pvcName)
+					return nil
```

Comment:
Same here - where there are actually errors? I understand why you'd return nil in the case where PV actually found

Reply:
My thinking:
1) Failure to get podList / pvc while node teardown (where a lot of things are happening) could be transient. We wouldn't want the node csi pod to terminate early because of a transient error
2) If a user is using graceful termination, they're looking for a delayed teardown behavior to let the application pods get deleted. We know this behavior is only going to happen when the node is being torn down, so a delayed teardown of the node csi pod in favor of letting application pods be deleted makes sense to me. Returning error would make the prestop hook fast fail -> would unblock node csi termination

Reply:
seems like ebs team takes the approach to fast fail, doesn't seem like a one way door so made the change.